src/Http/Http/src/StreamPipeReader.cs

```diff
@@ -0,0 +1,339 @@
+                _readHead = _readTail;
+                returnEnd = _readTail;
+                _readIndex = 0;
+            }
+            else if (consumedIndex == returnEnd.Length)
+            {
+                var nextBlock = returnEnd.NextSegment;
+                _readHead = nextBlock;
+                _readIndex = 0;
+                returnEnd = nextBlock;
+            }
+            else
+            {
+                _readHead = consumedSegment;
+                _readIndex = consumedIndex;
+            }
+
+            // Remove all blocks that are freed (except the last one)
+            while (returnStart != null && returnStart != returnEnd)
+            {
+                returnStart.ResetMemory();
+                returnStart = returnStart.NextSegment;
+            }
+        }
+
+        /// <inheritdoc />
+        public override void CancelPendingRead()
+        {
+            InternalTokenSource.Cancel();
+        }
+
+        /// <inheritdoc />
+        public override void Complete(Exception exception = null)
+        {
+            if (_isCompleted)
+            {
+                return;
+            }
+
+            _isCompleted = true;
+            if (exception != null)
+            {
+                _exceptionInfo = ExceptionDispatchInfo.Capture(exception);
+            }
+
+            var segment = _readHead;
+            while (segment != null)
+            {
+                segment.ResetMemory();
+                segment = segment.NextSegment;
+            }
+        }
+
+        /// <inheritdoc />
+        public override void OnWriterCompleted(Action<Exception, object> callback, object state)
+        {
+            throw new NotSupportedException("OnWriterCompleted is not supported");
+        }
+
+        /// <inheritdoc />
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
+
+            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
+            var tokenSource = InternalTokenSource;
+            if (TryReadInternal(tokenSource, out var readResult))
+            {
+                return readResult;
+            }
+
+            var reg = new CancellationTokenRegistration();
+            if (cancellationToken.CanBeCanceled)
+            {
+                reg = cancellationToken.Register(state => ((StreamPipeReader)state).Cancel(), this);
+            }
+
+            using (reg)
+            {
+                var isCanceled = false;
+                try
+                {
+                    AllocateReadTail();
+#if NETCOREAPP2_2
+                    var length = await _readingStream.ReadAsync(_readTail.AvailableMemory.Slice(_readTail.End), tokenSource.Token);
+#elif NETSTANDARD2_0
+                    if (!MemoryMarshal.TryGetArray<byte>(_readTail.AvailableMemory.Slice(_readTail.End), out var arraySegment))
```

Comment:
Extract `_readTail.AvailableMemory.Slice(_readTail.End)` and fallback to rented memory pool block (or `.ToArray()` when `TryGet` fails)

Reply:
Actually, I don't see how TryGet can fail here (as we control AvailableMemory).

Reply:
`IMemoryPool` controls `AvailibleMemory`

Reply:
Talked offline, it's better to throw here as every time someone wanted to read, it wouldn't allocate from the provided/default memory pool.